relayer/chain/ethereum/listener.go

```diff
@@ -107,23 +110,47 @@ func (li *Listener) pollEventsAndHeaders(ctx context.Context) error {
 				li.messages <- *msg
 			}
 		case gethheader := <-headers:
+			blockNumber := gethheader.Number.Uint64()
+			newestBlock = blockNumber
 			li.log.WithFields(logrus.Fields{
-				"blockNumber": gethheader.Number,
-			}).Info("Witnessed block header")
-
-			header, err := MakeHeaderFromEthHeader(gethheader, li.log)
+				"blockNumber": blockNumber,
+			}).Info("Witnessed new block header")
+			if blockNumber <= currentBlock+1 {
```

Comment:
I guess to enforce ordered submission of blocks we want to have a strict equality check here, right?

```suggestion
			if blockNumber == currentBlock+1 {
```

Reply:
The `<=` is intentional because it means that `newestBlock` and `currentBlock` rewind when we hit a block from a slower fork. If we then subsequently encounter a block at, say, `currentBlock+2` from a faster fork, this case will ignore it, but the `default` case will re-request `currentBlock+1` and `currentBlock+2`. This... works.... sort of. But I'm sure this logic can be thwarted. At the very least, it's really tough to reason about. The benefit is that it allows us to combine logic for sync-forwarding an non-sync-forward (i.e. when the Substrate node is caught up). But I'm going to rewrite all of this later

Reply:
Okay, understood. Would be great if we could discuss the architecture of this before you rewrite it. Thanks!